kolibri/auth/permissions/base.py

```diff
@@ -0,0 +1,272 @@
+
+    def user_can_update_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to update the provided <obj>."""
+        raise NotImplementedError("Override `user_can_update_object` in your permission class before you use it.")
+
+    def user_can_delete_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to delete the provided <obj>."""
+        raise NotImplementedError("Override `user_can_delete_object` in your permission class before you use it.")
+
+    def readable_by_user_filter(self, user, queryset):
+        """Applies a filter to the provided queryset, only returning items for which the user has read permission."""
+        raise NotImplementedError("Override `readable_by_user_filter` in your permission class before you use it.")
+
+    def __or__(self, other):
+        """
+        Allow two instances of BasePermission to be joined together with "|", which returns a permissions class
+        that grants permission for an object when *either* of the instances would grant permission for that object.
+        """
+        return PermissionsFromAny(self, other)
+
+    def __and__(self, other):
+        """
+        Allow two instances of BasePermission to be joined together with "&", which returns a permissions class
+        that grants permission for an object when *both* of the instances grant permission for that object.
+        """
+        return PermissionsFromAll(self, other)
+
+
+class RoleBasedPermissions(BasePermissions):
+    """
+    Permissions class that defines a requesting user's permissions in terms of his or her kinds of roles with respect
+    to a User or Collection related to the object.
+
+    To use this class, either subclass it and define the relevant attributes, or directly create an instance and pass
+    in the attributes as parameters.
+
+    See the comments below adjacent to the default attributes, for reference as to how to configure this class.
+    """
+
+    # For each crud operation, specify a list of role kinds that should have permission to perform that operation.
+    # By default, give permissions to no roles; override this or pass it into __init__.
+    can_be_created_by = None
+    can_be_read_by = None
+    can_be_updated_by = None
+    can_be_deleted_by = None
+
+    # Specify the field through which the role target (user or collection) will be referenced (or "." if the object itself
+    # is the target). The referenced field should be a ForeignKey to either the FacilityUser or the Collection model.
+    target_field = "user"  # by default, the target is the FacilityUser that is pointed to by the object's "user" field
+
+    def __init__(self, can_be_created_by=None, can_be_read_by=None, can_be_updated_by=None, can_be_deleted_by=None,
+                 can_be_written_by=None, target_field=None):
+
+        if can_be_created_by is not None:
+            self.can_be_created_by = can_be_created_by
```

Comment:
In this case does `self.can_be_created_by` refer to the class attribute above, or does it create a new instance attribute and assign it? For me it would be much clearer if setting the default values like this didn't rely on the instance attribute name shadowing the class attribute. Maybe the name shadowing isn't even necessary since all of `__init__`'s kwargs have default values anyway.